### DISCUSSION TITLE
[python] gramine-manifest ldd IndexError: list index out of range - How to use {{ ldd() }}

### DISCUSSION THREAD

**Discussion by Jonas-Eppard** [gramineproject/gramine#1410] [Q&A]
I want to run a C application inside gramine. The C application is linked against shared libraries. As described in the [docs](https://gramine.readthedocs.io/en/stable/manpages/gramine-manifest.html#ldd) I should be able to use ldd to get all the needed libraries for the application. But gramine-manifest throws an error if {{ ldd(...) }} is specified.

### Recreate
**Dockerfile**
```docker
FROM ubuntu:20.04

WORKDIR /enclave

# Install Gramine
RUN apt-get update \
    && rm -f /etc/ssl/certs/ca-bundle.crt \
    && apt-get reinstall ca-certificates --no-install-recommends -y \
    && update-ca-certificates \
    && apt-get install curl --no-install-recommends -y \
    && curl -fsSLo /usr/share/keyrings/gramine-keyring.gpg https://packages.gramineproject.io/gramine-keyring.gpg \
    && echo "deb [arch=amd64 signed-by=/usr/share/keyrings/gramine-keyring.gpg] https://packages.gramineproject.io/ focal main" \
        | tee /etc/apt/sources.list.d/gramine.list \
    && curl -fsSLo /usr/share/keyrings/intel-sgx-deb.asc https://download.01.org/intel-sgx/sgx_repo/ubuntu/intel-sgx-deb.key \
    && echo "deb [arch=amd64 signed-by=/usr/share/keyrings/intel-sgx-deb.asc] https://download.01.org/intel-sgx/sgx_repo/ubuntu focal main" \
        | tee /etc/apt/sources.list.d/intel-sgx.list \
    && apt-get update \
    && apt-get install gramine build-essential --no-install-recommends -y \
    && gramine-sgx-gen-private-key

RUN echo '#include <stdio.h>' > t.c \
    && echo 'int main(int argc, char **argv){printf("Hi");return 0;}' >> t.c \
    && gcc t.c -o t \
    && echo "sgx.trusted_files = [" > t.manifest.template \
    && echo '    "file:t"' >> t.manifest.template \
    && echo '    "file:{{ ldd('\'t\'')}}"' >> t.manifest.template \
    && echo ']' >> t.manifest.template

RUN gramine-manifest t.manifest.template > t.manifest
```
**Error**
This Docker file fails on build with: 
```
Traceback (most recent call last):
  File "/usr/bin/gramine-manifest", line 33, in <module>
    main() # pylint: disable=no-value-for-parameter
  File "/usr/lib/python3/dist-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/bin/gramine-manifest", line 29, in main
    manifest = Manifest.from_template(template, define)
  File "/usr/lib/python3/dist-packages/graminelibos/manifest.py", line 147, in from_template
    return cls(_env.from_string(template).render(**(variables or {})))
  File "/usr/lib/python3/dist-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/lib/python3/dist-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 3, in <module>
  File "/usr/lib/python3/dist-packages/graminelibos/gen_jinja_env.py", line 26, in ldd
    if line[1] == '=>' and line[2].startswith('/'):
IndexError: list index out of range
```
**t.c**
```c
#include <stdio.h>
int main(int argc, char **argv) {printf("Hi");return 0;}
```
**t.manifest.template**
```
sgx.trusted_files = [
    "file:t"
    "file:{{ ldd('t')}}"
]
```

**Output of ldd t**
```
        linux-vdso.so.1 (0x00007ffd56342000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f3a87190000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f3a8738c000)
```

I know that the manifest.template is not sufficient for a full enclaved application but it is a minimal failing example.

Is the usage of ldd wrong or did I encounter a bug?
Furthermore will it be sufficient to specify ldd in the trusted_files or should this also be inside of the fs.mounts?

**Comment by dimakuv** [comment]
You're probably the first person to use the `ldd()` syntax :)

@woju You introduced this `ldd()` syntax to `gramine-manifest`. Looks like it's not even our code, but Jinja2 complaining about the syntax? What's wrong with the syntax `{{ ldd('t')}`?

@Jonas-Eppard By the way, your way of using this `ldd` construct is wrong. I think `ldd()` expands to a list of paths, so it should be used similarly to this: https://github.com/gramineproject/gramine/blob/fdd3417b490ea0c287fb967432509b09161eb7f3/CI-Examples/python/python.manifest.template#L23-L25

**Reply by woju** [reply]
> Looks like it's not even our code, but Jinja2 complaining about the syntax?

No, it's in parsing of line.

> What's wrong with the syntax `{{ ldd('t')}`?

Nothing in Jinja, but the result is a list, so you'll get:

```yaml
sgx.trusted_files = [
    "file:t"
    "file:['lib1.so', 'lib2.so']"
]
```

TOML will parse this fine (apart from missing comma after `"file:t"`), but Gramine will think the filename is this string that starts with `[` and ends with `]`. Not what you wanted.

**Reply by Jonas-Eppard** [reply]
Thanks for the quick answer. You are right that it should follow the example you mentioned. 

I traced the error and took a look at the source. The error is during the evaluation of the ldd system command in this function:
```python
def ldd(*args):
    '''
    Args:
        binaries for which to generate manifest trusted files list.
    '''
    # Be careful: We have to skip vdso, which doesn't have a corresponding file on the disk (we
    # assume that such files have paths starting with '/', seems ldd always prints absolute paths).
    # Also, old ldd (from Ubuntu 16.04) prints vdso differently than newer ones:
    # old:
    #     linux-vdso.so.1 =>  (0x00007ffd31fee000)
    # new:
    #     linux-vdso.so.1 (0x00007ffd31fee000)
    ret = set()
    for line in subprocess.check_output(['ldd', *(os.fspath(i) for i in args)]).decode('ascii'):
        line = line.strip().split()
        if line[1] == '=>' and line[2].startswith('/'):
            ret.add(line[2])
        elif line[0].startswith('/') and line[1].startswith('/'):
            ret.add(line[0])
    return sorted(ret)
```
I went ahead and added some print statements:
```python
def ldd(*args):
    '''
    Args:
        binaries for which to generate manifest trusted files list.
    '''
    # Be careful: We have to skip vdso, which doesn't have a corresponding file on the disk (we
    # assume that such files have paths starting with '/', seems ldd always prints absolute paths).
    # Also, old ldd (from Ubuntu 16.04) prints vdso differently than newer ones:
    # old:
    #     linux-vdso.so.1 =>  (0x00007ffd31fee000)
    # new:
    #     linux-vdso.so.1 (0x00007ffd31fee000)
    ret = set()
    print(['ldd', *(os.fspath(i) for i in args)])
    print(subprocess.check_output(['ldd', *(os.fspath(i) for i in args)]).decode('ascii'))
    for line in subprocess.check_output(['ldd', *(os.fspath(i) for i in args)]).decode('ascii'):
        print("Line start of loop: '", line,"'", sep="")
        line = line.strip().split()
        print("Line after split: '", line, "'", sep="")
        if line[1] == '=>' and line[2].startswith('/'):
            ret.add(line[2])
        elif line[0].startswith('/') and line[1].startswith('/'):
            ret.add(line[0])
    return sorted(ret)
```
This gives inside into the problem. The invocation is correct but the iterator does not iterate correctly over the lines...
Output:
```
['ldd', 't']
        linux-vdso.so.1 (0x00007ffd56342000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f3a87190000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f3a8738c000)

Line start of loop: '   '
Line after split: '[]'
Traceback (most recent call last):
...
<same error as above>
...
IndexError: list index out of range
```

If i add a `.split("\n")` after the `.decode('ascii')` the iteration is corrected.
Output:
```
Line before split: '    linux-vdso.so.1 (0x00007ffc193ab000)'
Line before split: '    libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f762b577000)'

Line before split: '    /lib64/ld-linux-x86-64.so.2 (0x00007f762b77d000)'
Line before split: ''
```
But the last iteration still causes errors, because it will be an empty array...

Therefore I added a `.strip()` here: `.decode('ascii')|.split('\n'):`

This solves the error in Jinja2.

This is still not solved because if multiple executables are provided as specified in the [docs](https://gramine.readthedocs.io/en/stable/manpages/gramine-manifest.html)
Since the output of ldd contains seperator lines if called with multiple executables:

```
$ ldd t main
t:
    <lines for t>
main:
    <lines for main>
```

This would cause the lines `t:` or `main:` to also be evaluated in the loop body, but these lines only have 1 element so line[1] is out of bounds.

A possible solution would be to add a branch if the length of the split line is to short to continue. This also catches the case with the trailing new line, so even the `.strip()` in the loop header is not needed.

To sum up all my findings can be consolidated into the following diff:

```diff
     # new:
     #     linux-vdso.so.1 (0x00007ffd31fee000)
     ret = set()
-    for line in subprocess.check_output(['ldd', *(os.fspath(i) for i in args)]).decode('ascii'):
+    for line in subprocess.check_output(['ldd', *(os.fspath(i) for i in args)]).decode('ascii').split('\n'):
+        # print("Line before split: '", line, "'", sep="")
         line = line.strip().split()
-        if line[1] == '=>' and line[2].startswith('/'):
+        if len(line) < 2:
+            continue
+        elif line[1] == '=>' and line[2].startswith('/'):
             ret.add(line[2])
         elif line[0].startswith('/') and line[1].startswith('/'):
             ret.add(line[0])
```
This fixes the error I am getting in the previous mentioned docker file. I still have to apply the suggestion of @dimakuv but Jinji2 is working! 

TL;DR: There is a problem with Jinja2 but the above diff should resolve it

**Reply by dimakuv** [reply]
@Jonas-Eppard May I just say that this is an amazing write-up! Thanks.

**Comment by woju** [comment]
I'm attempting a fix in #1411, it's now running CI. Can I ask for `ldd --version`?

**Reply by Jonas-Eppard** [reply]
Inside the docker container the result of `ldd --version` is:
```
ldd (Ubuntu GLIBC 2.31-0ubuntu9.9) 2.31
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```
